src/zero/jdbc.clj

```diff
@@ -0,0 +1,270 @@
+  transaction or not (default true).
+
+  If the :multi? option is false (the default), the SQL statement should be
+  followed by the parameters for that statement.
+
+  If the :multi? option is true, the SQL statement should be followed by one or
+  more vectors of parameters, one for each application of the SQL statement.
+
+  If :return-keys is provided, db-do-prepared-return-keys will be called
+  instead of db-do-prepared, and the result will be a sequence of maps
+  containing the generated keys. If present, :row-fn will be applied. If :multi?
+  then :result-set-fn will also be applied if present. :as-arrays? may also be
+  specified (which will affect what :result-set-fn is passed).
+
+  If there are no parameters specified, executeUpdate will be used, otherwise
+  executeBatch will be used. This may affect what SQL you can run via execute!"
+  ([db sql-params] `(do
+                      (log/info "JDBC SQL execute! (without opts):" ~db ~sql-params)
+                      (jdbc/execute! ~db ~sql-params)))
+  ([db sql-params opts] `(do
+                           (log/info "JDBC SQL execute!:" ~db ~sql-params ~opts)
+                           (jdbc/execute! ~db ~sql-params ~opts))))
+
+(defmacro db-do-commands
+  "Logged alias for [[clojure.java.jdbc/db-do-commands]]:
+
+  Executes SQL commands on the specified database connection. Wraps the commands
+  in a transaction if transaction? is true. transaction? can be omitted and it
+  defaults to true. Accepts a single SQL command (string) or a vector of them.
+  Uses executeBatch. This may affect what SQL you can run via db-do-commands."
+  ([db sql-commands] `(do
+                        (log/info "JDBC SQL db-do-commands:" ~db ~sql-commands)
+                        (jdbc/db-do-commands ~db ~sql-commands)))
+  ([db transaction? sql-commands] `(do
+                                     (log/info "JDBC SQL db-do-commands:" ~db ~transaction? ~sql-commands)
+                                     (jdbc/db-do-commands ~db ~transaction? ~sql-commands))))
+
+(defmacro insert!
+  "Logged alias for [[clojure.java.jdbc/insert!]]:
+
+  Given a database connection, a table name and either a map representing a rows,
+  or a list of column names followed by a list of column values also representing
+  a single row, perform an insert.
+  When inserting a row as a map, the result is the database-specific form of the
+  generated keys, if available (note: PostgreSQL returns the whole row).
+  When inserting a row as a list of column values, the result is the count of
+  rows affected (1), if available (from getUpdateCount after executeBatch).
+  The row map or column value vector may be followed by a map of options:
+  The :transaction? option specifies whether to run in a transaction or not.
+  The default is true (use a transaction). The :entities option specifies how
+  to convert the table name and column names to SQL entities."
+  ([db table row] `(do
```

Comment:
Can you just foward the arguments?
```
(defn foo [& args]
     (log/info "called foo with arguments " args)
     (apply foo_impl args))
```

Reply:
Not in all cases, because macros can't be applied to. Capturing arguments manually also helps inform useful logging messages, especially for the `insert` and related functions. Capturing arguments manually also fills in useful tooltip information in IDEs.

A lot of what I did in this file was frankly to get around the fact that Clojure documentation can't link to other functions, which I feel is a really significant issue in the toolchain. Copying documentation and manually capturing arguments try to help get around the fact that a simple delegate pattern loses all context in Clojure. There's nothing I can do to change Clojure but putting in the work to manually capture arguments improves the tooltips and the logging.

I'm not worried about maintenance costs for this because clojure.java.jdbc changes would've broken our existing code, and here we at least now have a wrapper we can use to adapt.

Reply:
The ability to maintain a stable API is a really good point that i didn't think of. `clojure/jdbc` hasn't yet reached v1 yet so perhaps you're right.